### PR TITLE
feat(docs): replace gen_vimdoc.py with gen_vimdoc.lua

### PR DESCRIPTION
**Pull request by lewis6991**
### Problem

The documentation flow (`gen_vimdoc.py`) has several issues:
- it's not very versatile
- depends on doxygen
- doesn't work well with Lua code as it requires an awkward filter script to convert it into pseudo-C.
- The intermediate XML files and filters makes it too much like a rube goldberg machine.

### Solution

Re-implement the flow using Lua, LPEG and treesitter.

- `gen_vimdoc.py` is now replaced with `gen_vimdoc.lua` and replicates a portion of the logic.
- `lua2dox.lua` is gone!
- No more XML files.
- Doxygen is no longer used and instead we now use:
  - LPEG for comment parsing (see `scripts/luacats_grammar.lua` and `scripts/cdoc_grammar.lua`).
  - LPEG for C parsing (see `scripts/cdoc_parser.lua`)
  - Lua patterns for Lua parsing (see `scripts/luacats_parser.lua`).
  - Treesitter for Markdown parsing (see `scripts/text_utils.lua`).
- The generated `runtime/doc/*.mpack` files have been removed.
   - `scripts/gen_eval_files.lua` now instead uses `scripts/cdoc_parser.lua` directly.
- Text wrapping is implemented in `scripts/text_utils.lua` and appears to produce more consistent results (the main contributer to the diff of this change).

### Todo

<details>
<summary>Details</summary>

- [x] Good word wrapping
- [x] Parse lists:
  - Do we need a markdown parser?
     - ~https://github.com/jgm/lunamark: Comprehensive but large~
     - ~https://github.com/sebcat/lpeg-markdown: Too simple, doesn't handle nested lists.~
     - UPDATE: using the tree-sitter parser we bundle
- [x] Also make it work for C files.
- [x] Support custom LuaCATs extensions
  - [x] `@note`
  - [x] `@deprecated`
  - [x] `@defgroup` - REMOVED
  - [x] `@brief` 
  - [x] `@since`
- [x] Remove `gen_vimdoc.py`
- [x] Remove `lua2dox.lua`
- [x] Generate docs for Luadoc classes
   - Done but commented out in this PR.

</details>

### Future Work

- Lua objects documented with `@class` will now be easy to generate documentation for. The code for that is already included (but commented out) in this change.
  - This will additionally allow us to document the full types of function arguments by using `@field` instead of hand formatted text.
- Editorial changes to increase consistency with things like: help tags and function names. E.g. whether to include the `vim.` prefix or not.
- Switch the Lua parsing from Lua patterns to either a LPEG or Treesitter based parser.
- Support for more custom `@` annotations.
- Graduate the markdown renderer in `text_utils.lua` and use in LSP hover.

---

### Notes:
- The logic for generation function name/tags is still a bit of a mess. However now we have a bit more semantic information on the Lua files, this should now be easier to cleanup and create a more standard convention for documenting the various categories of functions.

- Before:
  ```
  $ wc -l scripts/gen_vimdoc.py scripts/lua2dox.lua
   1766 scripts/gen_vimdoc.py
    532 scripts/lua2dox.lua
   2298 total
  ```
  This PR:
  ```
  $ wc -l scripts/gen_vimdoc.lua scripts/luacats_parser.lua scripts/cdoc_grammar.lua scripts/cdoc_parser.lua scripts/text_utils.lua
    792 scripts/gen_vimdoc.lua
    521 scripts/luacats_parser.lua
     87 scripts/cdoc_grammar.lua
    223 scripts/cdoc_parser.lua
    242 scripts/text_utils.lua
   1865 total
  ```

### CODE REVIEW

File: runtime/doc/treesitter.txt [hunk 1]
```diff
@@ -945,7 +945,8 @@ add_directive({name}, {handler}, {force})
 
                                         *vim.treesitter.query.add_predicate()*
 add_predicate({name}, {handler}, {force})
-    Adds a new predicate to be used in queries
+    @field [integer] TSMetadata @field [string] integer|string Adds a new
+    predicate to be used in queries
```
Comment (gpanders) [inline runtime/doc/treesitter.txt:949]: 👀

File: runtime/doc/diagnostic.txt [hunk 2]
```diff
@@ -361,6 +361,144 @@ Example: >lua
...
+
+    Fields: ~
+        • {severity?} (`vim.diagnostic.SeverityFilter`)
+        • {priority?} (`integer`)
+        • {text?} (`table<vim.diagnostic.Severity,string>`)
+        • {numhl?} (`table<vim.diagnostic.Severity,string>`)
+        • {linehl?} (`table<vim.diagnostic.Severity,string>`)
+        • {texthl?} (`table<vim.diagnostic.Severity,string>`)
+
+*vim.diagnostic._extmark*
```
Comment (gpanders) [inline runtime/doc/diagnostic.txt:431]: Probably filter out classes beginning with `_`?

File: runtime/lua/vim/shared.lua [hunk 3]
```diff
@@ -64,7 +64,7 @@ function vim.deepcopy(orig, noref)
 end
 
 --- @class vim.gsplit.Opts
---- @field plain? boolean Use `sep` literally (as in string.find).
+--- @field plain? boolean Use `sep` literally (as in `|string.find|`).
```
Comment (gpanders) [inline runtime/lua/vim/shared.lua:67]:
```suggestion
--- @field plain? boolean Use `sep` literally (as in `|string.find()|`).
```

**Review by gpanders — COMMENTED** [review]
I know it's still marked as draft, just left a few comments on first pass.

**Comment by lewis6991** [comment]
No worries but this is incredibly rough.

It's not obvious, but the code here basically allows us to generate full docs for functions too, so I might pivot this to try and remove the doxygen middleman.

File: runtime/doc/treesitter.txt (hunk 1 above)
Comment (lewis6991) [reply runtime/doc/treesitter.txt:949]: No idea what happened here.

File: src/nvim/generators/luacats_grammar.lua
Comment (justinmk) [inline src/nvim/generators/luacats_grammar.lua:1]: Would it make sense to use this instead of our own LPEG grammar: https://github.com/tree-sitter-grammars/tree-sitter-luadoc

**Comment by lewis6991** [comment]
I'm really not sure treesitter is the right tool for this. Yes treesitter provides parsers, but it doesn't provide great APIs for what we're trying to achieve. The node structure, especially for C and Markdown, do not have a  an easy to follow hierarchy so there would need to be some normalisation step to put everything into a good shape.

Also note the Markdown parser doesn't have an active maintainer and is currently implemented as two separate parsers, so I'd be pretty nervous about depending on that.

I'm sure if enough time is put into it that we could make it work, but I feel it is reuse for the sake of reuse whilst making the stack a lot taller.

For C and Lua I don't think we really need a parser. Regexp should be good enough for parsing out prototypes. Otherwise a simple lpeg pattern can help us cover anything less trivial. It doesn't need to cover the whole language; just anything that follows a doc block. We also already depend on an lpeg grammar for the C API generators.

**Comment by justinmk** [comment]
> treesitter ... doesn't provide great APIs for what we're trying to achieve. The node structure, especially for C and Markdown, do not have a an easy to follow hierarchy ... Also note the Markdown parser doesn't have an active maintainer ...
>
> For C and Lua I don't think we really need a parser. Regexp should be good enough for parsing out prototypes. Otherwise a simple lpeg pattern can help us cover anything less trivial.

Interesting. Let's mention that in ~~luacats_parser.lua~~ `gen_vimdoc.lua` and the commit message, for future reference.

File: scripts/lua2vimdoc.lua
[the remark this answers is not in the document]
Comment (lewis6991) [reply scripts/lua2vimdoc.lua:1]: Sure as this develops, I'll rename this appropriately. Right now this is still driven through gen_vimdoc.py and is only used for Lua files.

**Comment by lewis6991** [comment]
Ok, so I've spent quite a bit more time on this and gone through a few rabbit holes.

Creating a simple markdown parser using LPEG isn't reasonably possible AFAIK since you need to somehow keep track of indentation for identifying blocks.

Next I tried looking at some other Markdown parsers implemented manually.
- First looked at https://github.com/mpeterv/markdown and it's _ok_, but wasn't particularly happy with its design.
- Then found https://github.com/frostming/marko, which is implemented in Python (very good Python I might add). So I tried to see if I could remove everything we didn't need and port to Lua. Turns out with just the block level stuff (list, paragraphs and code blocks) you're still left with 500-800 LOC. And that's with a good design.

**TLDR**

So I finally gave in and attempted to use the tree-sitter parser we bundle. I've got it to work pretty well, but it certainly has some challenges. One example being codeblocks seems to be parsed with incorrect ranges when in nested blocks like lists. It also doesn't correctly parse some stuff because of indentation in places.

I've been able to generate `diagnostic.txt` 95% the same to how it was with `gen_vimdoc.py` which I'm pretty happy with so this seems like a good route.

**Comment by clason** [comment]
> One example being codeblocks seems to be parsed with incorrect ranges when in nested blocks like lists.

We own the markdown parser now, so it should be possible to get fixes in relatively easily, if this helps.

File: runtime/doc/treesitter.txt [hunk 4]
```diff
@@ -880,8 +878,8 @@ start({bufnr}, {lang})                                *vim.treesitter.start()*
     Can be used in an ftplugin or FileType autocommand.
 
     Note: By default, disables regex syntax highlighting, which may be
-    required for some plugins. In this case, add `vim.bo.syntax = 'on'` after
-    the call to `start`.
+    required for some plugins. In this case, add ``vim.bo.syntax = 'on'``
```
Comment (clason) [inline runtime/doc/treesitter.txt:881]: 👀

Comment (lewis6991) [reply runtime/doc/treesitter.txt:881]: This is what it is in `treesitter.lua`. My new scripts have taken it verbatim.

File: runtime/doc/lua.txt [hunk 5]
```diff
@@ -2632,9 +2647,9 @@ vim.filetype.add({filetypes})                             *vim.filetype.add()*
             ['/etc/foo/config'] = 'toml',
           },
           pattern = {
-            ['.*&zwj;/etc/foo/.*'] = 'fooscript',
```
Comment (lewis6991) [inline runtime/doc/lua.txt:2635]: Was this intentional or was this doxygen bs?

File: runtime/doc/treesitter.txt [hunk 6]
```diff
@@ -1363,7 +1360,7 @@ LanguageTree:register_cbs({cbs}, {recursive})
       • {cbs}        (`table`) An |nvim_buf_attach()|-like table argument with
                      the following handlers:
                      • `on_bytes` : see |nvim_buf_attach()|, but this will be
-                       called after the parsers callback.
+                       called _after_ the parsers callback.
```
Comment (lewis6991) [inline runtime/doc/treesitter.txt:1363]: Do we want to keep this or remove the `_`?

Comment (clason) [reply runtime/doc/treesitter.txt:1363]: Remove, please. Vimdoc has no (text) markup.

Comment (clason) [reply runtime/doc/treesitter.txt:1363]: If we want emphasis, we should document this (as a "Neovimdoc" addition -- e.g., all uppercase). Then we can add treesitter support for it (e.g., make it bold).

Comment (lewis6991) [reply runtime/doc/treesitter.txt:1363]: There's so few of these. It's probably easier if I just remove these from the source.

Comment (clason) [reply runtime/doc/treesitter.txt:1363]: Agreed.

**Comment by lewis6991** [comment]
All the Lua files are now done. Feel free to review.

Now to tackle C...

File: runtime/lua/vim/_options.lua [hunk 7]
```diff
@@ -1,99 +1,96 @@
...
 ---
----@brief Nvim Lua provides an interface or "bridge" to Vimscript variables and
----functions, and editor commands and options.
----
----Objects passed over this bridge are COPIED (marshalled): there are no
----"references". |lua-guide-variables| For example, using \`vim.fn.remove()\` on
----a Lua list copies the list object to Vimscript and does NOT modify the Lua
----list:
+--- Objects passed over this bridge are COPIED (marshalled): there are no
+--- "references". |lua-guide-variables| For example, using \`vim.fn.remove()\` on
```
Comment (zeertzjq) [inline runtime/lua/vim/_options.lua:5]:
```suggestion
--- "references". |lua-guide-variables| For example, using `vim.fn.remove()` on
```

File: runtime/doc/lua.txt [hunk 8]
```diff
@@ -1148,20 +1152,17 @@ vim.wait({time}, {callback}, {interval}, {fast_only})             *vim.wait()*
 ==============================================================================
 LUA-VIMSCRIPT BRIDGE                                           *lua-vimscript*
 
-
-Nvim Lua provides an interface or "bridge" to Vimscript variables and
-functions, and editor commands and options.
+Nvim Lua provides an interface or "bridge" to Vimscript variables
+andfunctions, and editor commands and options.
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:1156]: Missing space

File: runtime/doc/lua.txt [hunk 9]
```diff
@@ -3008,6 +3019,40 @@ The LPeg library for parsing expression grammars is included as `vim.lpeg`
...
+    Parameters: ~
+      • {subject}  (`string`)
+      • {init}     (`integer?`)
+
+    Return: ~
+        (`integer|vim.lpeg.Capture?`)
+
+vim.lpeg                                                            *vim.lpeg*
+    *LPeg* is a new pattern-matching library for Lua, based on [Parsing
+    Expression Grammars](https://bford.info/packrat/) (PEGs).
```
Comment (clason) [inline runtime/doc/lua.txt:3054]: Duplicate tag -- and section (see intro).

Comment (clason) [reply runtime/doc/lua.txt:3054]: (And again, faux markup in the docs, which needs to go.)

File: runtime/lua/vim/_meta/lpeg.lua [hunk 10]
```diff
@@ -5,20 +5,19 @@ error('Cannot require a meta file')
...
----Parsing Expression Grammars (https://bford.info/packrat/) (PEGs).
+--- @brief <pre>help
+--- LPeg is a pattern-matching library for Lua, based on
+--- Parsing Expression Grammars (https://bford.info/packrat/) (PEGs).
 ---
----                                                                    *lua-lpeg*
----                                                            *vim.lpeg.Pattern*
----The LPeg library for parsing expression grammars is included as `vim.lpeg`
----(https://www.inf.puc-rio.br/~roberto/lpeg/).
+---                                                                     *lua-lpeg*
```
Comment (clason) [inline runtime/lua/vim/_meta/lpeg.lua:12]: This tag seems to have vanished? (It's used in the docs.)

File: runtime/lua/vim/_meta/lpeg.lua [hunk 11]
```diff
@@ -5,20 +5,19 @@ error('Cannot require a meta file')
...
 ---
----In addition, its regex-like interface is available as |vim.re|
----(https://www.inf.puc-rio.br/~roberto/lpeg/re.html).
+--- In addition, its regex-like interface is available as |vim.re|
+--- (https://www.inf.puc-rio.br/~roberto/lpeg/re.html).
 ---
----</pre>
+--- </pre>
 
 --- *LPeg* is a new pattern-matching library for Lua, based on [Parsing Expression Grammars](https://bford.info/packrat/) (PEGs).
```
Comment (clason) [inline runtime/lua/vim/_meta/lpeg.lua:22]:
```suggestion
```

Redundant, uses faux markup, and trips up the docgen, so it needs to go.

**Comment by lewis6991** [comment]
Finished C. 

Negative diffstat achieved!

Now need to fix everything I broke. i.e. anything that depended on the mpack files outputted from `gen_vimdoc.py`.

File: .github/workflows/docs.yml [hunk 12]
```diff
@@ -20,10 +20,6 @@ jobs:
       - uses: actions/checkout@v4
       - uses: ./.github/actions/setup
 
-      - name: Install dependencies
-        run: |
-          sudo apt-get install -y doxygen python3-msgpack
```
Comment (clason) [inline .github/workflows/docs.yml:25]: 🔥

File: runtime/lua/vim/_meta/lpeg.lua (hunk 11 above)
Comment (lewis6991) [reply runtime/lua/vim/_meta/lpeg.lua:22]: This removes the the *vim.lpeg* tag and specifically the text:

```
vim.lpeg                                                            *vim.lpeg*
    *LPeg* is a new pattern-matching library for Lua, based on [Parsing
    Expression Grammars](https://bford.info/packrat/) (PEGs).
```

is this still ok the remove?

Comment (clason) [reply runtime/lua/vim/_meta/lpeg.lua:22]: There's already a `vim.lpeg` tag above this; I think this is why CI was failing before. And, yes, the text _should_ go: it duplicates the section header.

Comment (clason) [reply runtime/lua/vim/_meta/lpeg.lua:22]: Of course, you could also restructure the section so that the header is not static but generated from this text, but that sounds like more work.

File: runtime/lua/vim/_meta/lpeg.lua (hunk 10 above)
Comment (lewis6991) [reply runtime/lua/vim/_meta/lpeg.lua:12]: It's still there from what I can tell. `lua.txt:3014`

File: runtime/lua/vim/_meta/lpeg.lua (hunk 11 above)
Comment (lewis6991) [reply runtime/lua/vim/_meta/lpeg.lua]: Ok, removed and added the tag above.

File: runtime/lua/vim/_meta/lpeg.lua [hunk 13]
```diff
@@ -5,22 +5,21 @@ error('Cannot require a meta file')
...
+--- LPeg is a pattern-matching library for Lua, based on
+--- Parsing Expression Grammars (https://bford.info/packrat/) (PEGs).
 ---
----                                                                    *lua-lpeg*
----                                                            *vim.lpeg.Pattern*
----The LPeg library for parsing expression grammars is included as `vim.lpeg`
----(https://www.inf.puc-rio.br/~roberto/lpeg/).
+---                                                                     *lua-lpeg*
+---                                                             *vim.lpeg.Pattern*
+---                                                                     *vim.lpeg*
```
Comment (clason) [inline runtime/lua/vim/_meta/lpeg.lua:14]: Duplicate tag (manually set at the top of the section).

File: scripts/gen_vimdoc.lua [hunk 14]
```diff
@@ -0,0 +1,1038 @@
+--- Generates Nvim :help docs from Lua docstrings
```
Comment (clason) [inline scripts/gen_vimdoc.lua:1]:
```suggestion
#!/usr/bin/env -S nvim -l

--- Generates Nvim :help docs from Lua docstrings
```
and `chmod +x`?

File: runtime/lua/vim/shared.lua [hunk 15]
```diff
@@ -260,7 +261,7 @@ end
 ---@generic T
 ---@param func fun(value: T): boolean (function) Function
 ---@param t table<any, T> (table) Table
----@return T[] (table) Table of filtered values
+---@return T[] : Table of filtered values
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/lua/vim/shared.lua:264]: Yeah, I can't remember which one is preferred.

[the remark this answers is not in the document]
Comment (MariaSolOs) [reply runtime/lua/vim/shared.lua:264]: I don't this there's a preference, I was just curious.

File: runtime/doc/treesitter.txt [hunk 16]
```diff
@@ -933,6 +933,8 @@ get_filetypes({lang})                *vim.treesitter.language.get_filetypes()*
         (`string[]`) filetypes
 
 get_lang({filetype})                      *vim.treesitter.language.get_lang()*
+    TODO: Documentation
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/treesitter.txt:936]: This was added (inconsistently) by gen_vimdoc.py. I think it only applied it to the API functions. The new flow applies it to everything.

I can remove if we don't like it

[the remark this answers is not in the document]
Comment (MariaSolOs) [reply runtime/doc/treesitter.txt:936]: I personally don't think we should expose TODO comments in public documentation.

File: runtime/lua/vim/shared.lua (hunk 15 above)
[the remark this answers is not in the document]
Comment (clason) [reply runtime/lua/vim/shared.lua:264]: I like the colon...

File: scripts/text_utils.lua [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+  return extract(root) or {}
+end
+
+--- @param x string
+--- @param start_indent integer
+--- @param indent integer
+--- @param text_width integer
+--- @return string
+function M.wrap(x, start_indent, indent, text_width)
+  text_width = text_width or 78
```
Comment (MariaSolOs) [inline scripts/text_utils.lua:66]: Why 78?

File: scripts/text_utils.lua [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+    parts[#parts + 1] = M.wrap(text, start_indent, indent, text_width)
+    parts[#parts + 1] = '\n'
+  elseif node.type == 'code_fence_content' then
+    local lines = vim.split(node.text:gsub('\n%s*$', ''), '\n')
+
+    local cindent = indent + INDENTATION
+    if level > 3 then
+      -- The tree-sitter markdown parser doesn't parse the code blocks indents
+      -- correctly in lists. Fudge it!
+      lines[1] = '    ' .. lines[1] -- ¯\_(ツ)_/¯
```
Comment (MariaSolOs) [inline scripts/text_utils.lua:120]:
```suggestion
      lines[1] = (' '):rep(INDENTATION) .. lines[1] -- ¯\_(ツ)_/¯
```

File: scripts/text_utils.lua [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+        parts[#parts + 1] = string.rep(' ', cindent)
+        parts[#parts + 1] = l
+      end
+      parts[#parts + 1] = '\n'
+    end
+  elseif node.type == 'fenced_code_block' then
+    parts[#parts + 1] = string.rep(' ', indent)
+    parts[#parts + 1] = '>'
+    for _, child in ipairs(node) do
+      -- if child.type == 'info_string' then
```
Comment (MariaSolOs) [inline scripts/text_utils.lua:140]: To be removed?

File: scripts/text_utils.lua (hunk 17 above)
Comment (zeertzjq) [reply scripts/text_utils.lua:66]: Most help files use this width.

**Comment by clason** [comment]
just in time, too: https://github.com/Homebrew/homebrew-core/pull/164148

File: scripts/text_utils.lua (hunk 18 above)
Comment (lewis6991) [reply scripts/text_utils.lua:120]: I'm not even sure `INDENTATION` is even correct here, though it happens to be the same. This is a complete fudge that just minimizes the diff and seems to produce the most consistent results.

The real action here is to fix the range generation in the Markdown parser directly.

File: runtime/doc/lua.txt [hunk 20]
```diff
@@ -2916,8 +2932,8 @@ vim.fs.joinpath({...})                                     *vim.fs.joinpath()*
 vim.fs.normalize({path}, {opts})                          *vim.fs.normalize()*
     Normalize a path to a standard format. A tilde (~) character at the
     beginning of the path is expanded to the user's home directory and any
-    backslash (\) characters are converted to forward slashes (/). Environment
-    variables are also expanded.
+    backslash (\\) characters are converted to forward slashes (/).
```
Comment (zeertzjq) [inline runtime/doc/lua.txt:2935]: It seems that `fs.lua` needs a `\\` to `\` change.

File: runtime/doc/lua.txt [hunk 21]
```diff
@@ -1326,6 +1331,8 @@ In any of the above examples, to replicate the behavior |:setlocal|, use
 `vim.opt_local`. Additionally, to replicate the behavior of |:setglobal|, use
 `vim.opt_global`.
 
+
+
```
Comment (gpanders) [inline runtime/doc/lua.txt]: More extra newlines

File: runtime/doc/api.txt [hunk 22]
```diff
@@ -3682,14 +3663,13 @@ nvim_ui_pum_set_height({height})                    *nvim_ui_pum_set_height()*
     Parameters: ~
       • {height}  Popupmenu height, must be greater than zero.
 
-nvim_ui_set_focus({gained})                              *nvim_ui_set_focus()*
-    Tells the nvim server if focus was gained or lost by the GUI.
+nvim_ui_set_focus({gained}, {error})                     *nvim_ui_set_focus()*
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/api.txt:3665]: Nope. Good catch.

File: runtime/doc/lua.txt [hunk 23]
```diff
@@ -884,6 +885,9 @@ Log levels are one of the values defined in `vim.log.levels`:
     vim.log.levels.WARN
     vim.log.levels.OFF
 
+
+
+
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/lua.txt]: I've removed one of these lines. The other might need to stay.

File: runtime/doc/lua.txt [hunk 24]
```diff
@@ -2391,6 +2400,8 @@ vim.validate({opt})                                           *vim.validate()*
 ==============================================================================
 Lua module: vim.loader                                            *vim.loader*
 
+Loader:loadfile()                                          *Loader:loadfile()*
+
```
Comment (wookayin) [inline runtime/doc/lua.txt:2404]: `@nodoc?`

File: runtime/lua/vim/treesitter/query.lua [hunk 25]
```diff
@@ -231,7 +231,7 @@ end
 ---@param lang string Language to use for the query
 ---@param query_name string Name of the query (e.g. "highlights")
 ---
----@return vim.treesitter.Query|nil -- Parsed query. `nil` if no query files are found.
+---@return vim.treesitter.Query|nil Parsed query. `nil` if no query files are found.
```
[the remark this answers is not in the document]
Comment (wookayin) [reply runtime/lua/vim/treesitter/query.lua:234]:
```suggestion
---@return vim.treesitter.Query|nil : Parsed query. `nil` if no query files are found.
```

File: runtime/doc/lua.txt (hunk 23 above)
[the remark this answers is not in the document]
Comment (wookayin) [reply runtime/doc/lua.txt]: (duplicate to the above comments, please never mind)

File: runtime/doc/lsp.txt [hunk 26]
```diff
@@ -712,10 +712,9 @@ buf_request_sync({bufnr}, {method}, {params}, {timeout_ms})
         (`string?`) err On timeout, cancel, or error, `err` is a string
         describing the failure reason, and `result` is nil.
 
-client()                                                      *vim.lsp.client*
+client                                                        *vim.lsp.client*
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/lsp.txt:715]: I don't think it is a function. I believe this was just used so we can document the properties of the client.

File: runtime/doc/lua.txt (hunk 24 above)
Comment (lewis6991) [reply runtime/doc/lua.txt:2404]: It's already got `@private`. The issue is it has `-- luacheck: ignore 312`, and the `--` breaks away the doc block from the function.

File: runtime/doc/lsp.txt (hunk 26 above)
[the remark this answers is not in the document]
Comment (clason) [reply runtime/doc/lsp.txt:715]: It is not a function; the very next line says "LSP client object." (Note that the tag also doesn't include parens.)

File: runtime/doc/lua.txt [hunk 27]
```diff
@@ -3008,6 +3026,36 @@ The LPeg library for parsing expression grammars is included as `vim.lpeg`
 In addition, its regex-like interface is available as |vim.re|
 (https://www.inf.puc-rio.br/~roberto/lpeg/re.html).
 
+
+
+
+Pattern:match({subject}, {init})                             *Pattern:match()*
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/lua.txt:3026]: Yes, getting all the tags to match was a pain.

Instead I've gone with a slightly more generic scheme:

```lua
--- @param fun nvim.luacats.parser.fun
--- @return string
local function fn_helptag_fmt_common(fun)
  local fn_sfx = fun.table and '' or '()'
  if fun.classvar then
    return fmt('*%s:%s%s*', fun.classvar, fun.name, fn_sfx)
  end
  if fun.module then
    return fmt('*%s.%s%s*', fun.module, fun.name, fn_sfx)
  end
  return fmt('*%s%s*', fun.name, fn_sfx)
end
```

Here we are documenting class/object methods as `*Class:method()*` which is generally consistent with what we do with other files.

File: runtime/lua/vim/treesitter/query.lua (hunk 25 above)
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/lua/vim/treesitter/query.lua:234]: We support the delimiter. I've added it.

```lua
  at_return = Ct(
    Cg(P('return'), 'kind')
      * ws
      * parenOpt(comma(Ct(Cg(v.ltype, 'type') * opt(ws * Cg(ident, 'name')))))
      * opt(desc_delim * Cg(rep(any), 'desc'))
  ),
```

File: runtime/doc/lua.txt [hunk 28]
```diff
@@ -3926,7 +3975,7 @@ Iter:fold({init}, {f})                                           *Iter:fold()*
 
     Parameters: ~
       • {init}  (`any`) Initial value of the accumulator.
-      • {f}     (`fun(acc:any, ...):any`) Accumulation function.
+      • {f}     (`fun(acc:A, ...):A`) Accumulation function.
```
[the remark this answers is not in the document]
Comment (lewis6991) [reply runtime/doc/lua.txt:3972]: I was going to leave this since I didn't think the extra complexity was worth normalizing this one occurance.

File: runtime/doc/lua.txt (hunk 24 above)
Comment (lewis6991) [reply runtime/doc/lua.txt]: Fixed now.

**Review by clason — APPROVED** [review]
Beautiful! Thanks for tackling this!

**Review by bfredl — APPROVED** [review]
Minus 1766 lines of python. I approve.

**Review by gpanders — APPROVED** [review]
The approval train is leaving the station 🚉

**Comment by josephjamesmorris700-hub** [comment]
Very confusing
